Chapter 24: Shaders

AristurtleDev opened the pull request:
Chapter 24: Shaders

This chapter gives an overview of shaders in MonoGame and implements a simple grayscale shader effect for when the game is paused or a game over state occurs to visually indicate that the game is inactive at that point.

Live Preview: https://aristurtle.net/tutorials/building_2d_games/24_shaders/

AristurtleDev committed:
Initial draft

AristurtleDev committed:
Add video showing effect

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+> [!IMPORTANT]
+> This chapter is an introduction to shaders in MonoGame and will focus on the basic foundation of understanding how to create shader effect (.fx) files, loading them through the content pipeline, and using them in your game.
+>
+> It is out of scope for this tutorial series to go into depth about the syntax and many built in functions for the shader language itself.  If you would like to dive deeper into learning the language itself, a good place to start would be the [Microsoft Learn: High-level shader language (HLSL)](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl) documentation.
```
The first paragraph is fine for stating the scope of the shader tutorial.

In the second, rather than highlighting it is "out of scope", maybe reword to say something like.

If you want to learn more about the shader language, a good place to start would be the...
To see just what can be done with shaders, check out ShaderToy here..

AristurtleDev replied:
Updated to state the following

> [!IMPORTANT]
> This chapter is an introduction to shaders in MonoGame and will focus on the basic foundation fo understanding how to create shader effect (*.fx*) files, loading them through the content pipeline, and using them in your game.
>
> If you want to learn more about the shader language itself, a good place to start would be the [High-level shader language](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl) documentation on Microsoft Learn.  
>
> For inspiration on what can be achieved with shaders, check out [ShaderToy](https://www.shadertoy.com), which showcases real-time shader effects created by others.  Note that ShaderToy uses OpenGL Shading Language (GLSL) which has some syntactic differences from HLSL, but the underlying concepts and mathematics are very similar for inspiration.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+Vertex shaders are especially important in 3D games, but they can also be used in 2D games for special effects that involve moving or distorting sprite positions.
+
+> [!NOTE]
+> Even in 2D games like ours, MonoGame actually draws sprites by mapping textures onto simple 3D shapes called quads (rectangles made of two triangles). Each quad has four vertices, one at each corner. While we don't normally think about these vertices in 2D development, they're still there behind the scenes, and vertex shaders process them before rendering.
```
You are killing it with these little call outs, pointing our that 2D is really just FLAT 3D :D

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+### Shader Languages and Cross-Platform Considerations
+
+MonoGame uses High-Level Shader Language (HLSL) for writing shader effects. HLSL is a C-like programming language developed by Microsoft for DirectX. However, since MonoGame supports multiple platforms, including those that use OpenGL instead of DirectX, it needs a way to make shaders work everywhere.
```
MonoGame uses "the" High-Level

Also for the words High-Level Shader Language, include a hyper link to the MS HLSL reference guide.

SimonDarksideJ replied:
"including those that use OpenGL instead of DirectX"
Consider pointing out that DirectX uses HLSL where as OpenGL uses GLSL.   This is pointed out in the next paragraph but should be here.

---
(suggestion)
As MonoGame also supports OpenGL which uses the [OpenGL Shading Language](https://www.khronos.org/opengl/wiki/OpenGL_Shading_Language) (GLSL) instead of DirectX, it needs a way to make shaders work everywhere.

AristurtleDev replied:
Section has been updated now to:

### Shader Languages and Cross-Platform Considerations

MonoGame uses the [High-Level Shader Language (HLSL)](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl) for writing shader effects. HLSL is a C-like programming language developed by Microsoft for DirectX. As MonoGame also supports OpenGL which uses the [OpenGL Shading Language (GLSL)](https://www.khronos.org/opengl/wiki/OpenGL_Shading_Language) instead of DirectX, it needs a way to make shaders work everywhere.

This is where MojoShader comes in. MojoShader is a library that automatically translates your HLSL shader code into whatever format the target platform requires (like GLSL for OpenGL platforms). This translation happens during the content build process when you compile your game.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+| `vs_2_0`            | `ps_2_0`           |
+| `vs_3_0`            | `ps_3_0`           |
+
+For maximum compatibility, it's best to target the lower shader models:
```
Can we also add an information note stating that MonoGame is currently planning to upgrade the graphics pipeline to support Vulkan and DX12, which will further increase the graphical support, stay tuned

Or do you think that would be too much?

AristurtleDev replied:
It's good information to include since it's a hot topic for people that want to work in the shader space.

Added it as a note as follows

> [!NOTE]
> MonoGame is currently planning to upgrade its graphics pipeline to support Vulkan and DirectX 12, which will significantly enhance graphical capabilities and shader support across platforms, enabling more advanced visual effects and better performance in future versions.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+When you create a new Sprite Effect (.fx) file in MonoGame using the MGCB Editor, it generates a default template file.  Let's example this template to understand the foundation we'll build upon:
+
+[!code-c[](./snippets/defaultshader.fx)]
```
Can we add a note about why this is important

```
#if OPENGL
    #define SV_POSITION POSITION
    #define VS_SHADERMODEL vs_3_0
    #define PS_SHADERMODEL ps_3_0
#else
    #define VS_SHADERMODEL vs_4_0_level_9_1
    #define PS_SHADERMODEL ps_4_0_level_9_1
#endif
```

Or even as a new section before this one, just to highlight what it is indeed doing and what constraints it adds.  I think it is critical to point out that developers need to constrain their shader language use to the minimum level to support both platforms, else they will need more of these sections if they want it more advanced in DX.

SimonDarksideJ replied:
I know you go into this with a single paragraph below, but I feel it's important to highlight how critical this block is and its impact on how you write the shader.

AristurtleDev replied:
Add the following admonition after that code section

 > [!IMPORTANT]
> These preprocessor directives ensure your shader works across different platforms by defining appropriate shader models and semantics.  When MonoGame compiles your shader:
>
> - On OpenGL platforms (macOS, Linux, etc.), it uses the `OPENGL` defiition, setting shader models to `vs_3_0` and `ps_3_0`.
> - On DirectX platforms (Windows, Xbox), it uses the DirectX path, setting shader models to `vs_4_0_level_9_1` and `ps_4_0_level_9_1`.
>
> This compatibility block directly impacts which HLSL features you can use in your shader.  To maintain cross-platform compatibility, you should restrict yourself to features available at the lowest shader model you target.  Using advanced features only available in higher shader models will require additional conditional compilation blocks and platform-specific code paths.

SimonDarksideJ replied:
Shader "Profiles", as noted in another comment!

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+    [!code-c[](./snippets/defaultshader.fx?start=10&end=10)]
+
+3. **Sampler state**: This creates a `sampler2D` called `SpriteTextureSampler` that is used sample the texture received from the [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch).  Think of this as similar to the [**SamplerState**](xref:Microsoft.Xna.Framework.Graphics.SamplerState) we discussed in [Chapter 18](../18_texture_sampling/index.md).
```
Need to keep the language simpler, maybe something like

> Sampler state: This creates a sampler2D called SpriteTextureSampler that is used sample the texture received from the [SpriteBatch](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch).  In essence it tells the graphics card how to read the texture data from the texture file/stream that it is being passed, similar to the [SamplerState](xref:Microsoft.Xna.Framework.Graphics.SamplerState) we discussed in [Chapter 18](https://aristurtle.net/tutorials/building_2d_games/18_texture_sampling/index.html).

AristurtleDev replied:
Reworded to the following

> 3. **Sampler state**: This creates a `sampler2D` called `SpriteTextureSampler` that controls how the shader reads pixel data from the texture received from the [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch).  It defines properties like [filtering](../18_texture_sampling/index.md#filtering-modes) and [addressing](../18_texture_sampling/index.md#addressing-modes) modes for texture sampling, similar to the [**SamplerState**](xref:Microsoft.Xna.Framework.Graphics.SamplerState) we discussed in [Chapter 18](../18_texture_sampling/index.md).

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+- Special transitions (fades, dissolves, color shifts)
+- Dynamic lighting effects
+
+For our Dungeon Slime game, we'll focus primarily on pixel shaders since we want to create a color effect for our game over state.
```
Add a note that Compute shaders, etc exist but are not supported by MonoGame yet and are not covered in this beginner tutorial

AristurtleDev replied:
Add this note

> [!NOTE]
> There are other types of shaders beyond vertex and pixel shaders, such as compute shaders, geometry shaders, and hull/domain shaders.  These more advanced shader types enabled powerful features like physics simulations, procedural geometry, and complex post-processing effects.  However, they are not currently supported in the standard MonoGame implementation and are beyond the scope of this beginner tutorial.  As the MonoGame graphics pipeline evolves, support for these advanced shader types may be added in future versions.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+    >
+    > For example, `SV_POSITION` tells the system "this contains the final screen position," `COLOR0` means "this contains color data," and `TEXCOORD0` means "this contains texture coordinates."
+    >
+    > Semantics are required in HLSL to connect your shader variables with the graphics pipeline.  Without them, the GPU would not know what each piece of data represents or how to use it correctly.
```
Include link to the MS docs on semantics
https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-semantics

Always include links to source references where external terms are used, where possible.

AristurtleDev replied:
Added link

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+Think of shaders as special instructions that tell your GPU exactly how to display each pixel or transform each vertex in your game. By customizing these instructions, you can create a wide variety of visual effects without having to change your original artwork.
+
+### Types of Shaders
```
I always feel when talking about shaders, a small diagram is needed to show the basic flow of:

* Sampler
* -> Vertex Shader
* -> Pixel Shader
* -> Output

AristurtleDev replied:
Created new section with diagram as follows

> ### The Shader Pipeline
> 
> To understand how shaders work, it helps to visualize how data flow through the rendering pipeline.
> 
> | ![shader-pipeline](https://github.com/user-attachments/assets/0c3fa269-3926-441b-9cd1-1a7c80619b0c) |
> | :----------------------------------------------------------------------------------------------------------------------: |
> |               **Figure 24-1: Basic shader pipeline showing how data flows through the rendering process**                |
> 
> This diagram illustrates the fundamental steps of the shader pipeline:
> 
> 1. **Input Data**: The process begins with input data that is sent to the GPU to render:
>    1. **Vertex Data**: The vertex data (positions, colors, etc.) that define the geometry of what is being drawn.
>    2. **Texture**: The image data that will be applied to the geometry.
> 2. **Vertex Shader**: Processes each vertex, calculating its final position on the screen and passing the data to the next stage.
> 3. **Sampler**: Controls how texture data is accessed, applying [filtering](../18_texture_sampling/index.md#filtering-modes) (how pixels blend when scaled) and [addressing](../18_texture_sampling/index.md#addressing-modes) (what happens at texture edges).
> 4. **Pixel Shader**: Takes the transformed vertices and sampled texture data to calculate the final color of each pixel.
> 5. **Output**: The final rendered image that appears on your screen.
> 
> When working with [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch) (2D) in MonoGame, the framework handles most of the vertex shader work automatically, which is why we will focus primarily on writing pixel shaders for visual effects.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+> [!NOTE]
+> You may have noticed that the default shader does not define a vertex shader method that would execute and return back the `VertexShaderOutput` value that is used as the input for the **Pixel Shader** function.
+>
+> When you're using [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch), MonoGame uses a built-in vertex shader under the hood that executes and creates this value to pass to the pixel shader function.
```
I actually did not know this :D

AristurtleDev replied:
Yep, it uses the [SpriteEffect.fx](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Platform/Graphics/Effect/Resources/SpriteEffect.fx)

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+Think of it like baking a cake:
+
+- The technique is the overall cake recipe.
```
Now I'm hungry, but a fantastic way to break down the explanations of techniques

You could include a reference to the Ship Game blur shader as a more advanced example of a shader with multiple techniques and passes, to blow their mind :P 
https://github.com/MonoGame/MonoGame.Samples/blob/3.8.2/ShipGame/ShipGame.Core/Content/shaders/Blur.fx

AristurtleDev replied:
Added the following tip

> [!TIP]
> For a real-world example of a blur shader with multiple techniques and passes, take a look at the [Blur shader from the MonoGme Ship Game](https://github.com/MonoGame/MonoGame.Samples/blob/3.8.2/ShipGame/ShipGame.Core/Content/shaders/Blur.fx) sample.  This shader demonstrates how visual effects can be built by combining multiple rendering passes, with separate horizontal and vertical blur passes that work together to create a final blur effect.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+## Using Effects With SpriteBatch
+
+Once you've loaded a shader effect, applying it to your game's visuals requires integrating it with the [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch). The effect is specified during the `Begin` call, but is actually applied during drawing operations or when `End` is called (depending on the SpriteSortMode).
```
You have!

AristurtleDev replied:
Contraction removed.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+## Loading Shader Effects
+
+Shader effects are loaded in MonoGame using the content manager, much like any other asset that is compiled using the content pipeline.  The type to use when loading a shader is [**Effect**](xref:Microsoft.Xna.Framework.Graphics.Effect).
```
Initial paragraph to explain this is only an example and we are not updating the game yet.

"Before we add a new sprite shader to our game, let us first walkthrough what is involved in using a shader in our SpriteBatch calls."

AristurtleDev replied:
I've updated this section slightly.

There is now a level 2 header section that intros into the loading and using sections which were moved to level 3 headings.  The intro section states

> Now that we understand what shaders are, we can explore how to integrate them into a MonoGame project. Before implementing a shader in our game, we will first take a look at the process of loading and using shaders with [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch).

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+## Creating a Grayscale Shader
+
+Now that we understand the basics of shaders, we will create a simple grayscale effect that we can apply when the game is paused or there is a game over to provide visual feedback to the player that the game is inactive.
```
"Back to building out our game and adding a cool effect.  

Now that we understand the basics of shaders, we will create a simple grayscale effect that we can apply when the game is paused or there is a game over to provide visual feedback to the player that the game is inactive."

Just making it clearer we are switching back into DOING mode from the theory above. (based on other reviewer statements)

AristurtleDev replied:
I have changed the heading of the section and slightly reworded the intro to be more clear about that

> ## Implementing a Shader In Our Game
> 
> Now it is time to implement a shader for our game.  The shader we will create is a simple grayscale effect that can be applied when the game is paused or there is a game over to provide visual feedback to the player that the game is inactive.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+The key modifications made to create this grayscale effect include:
+
+1. **Added a Parameter**: A `Saturation` parameter was added that controls he intensity fo the grayscale effect.  When set to 0, the image will be fully grayscale.  When set to 1, the image will be its original color.  Values in between create a partial grayscale effect.
```
"Added a Parameter: A Saturation parameter was added that controls he intensity for the grayscale effect:

* When set to 0 the image will be fully grayscale
* When set to 1 the image will be its original color.

 Values in-between create a partial grayscale effect."

AristurtleDev replied:
Updated to be a list

> 1. **Added a Parameter**: A `Saturation` parameter was added that controls he intensity of the grayscale effect.  
> 
>   - When set to 0, the image will be fully grayscale.  
>   - When set to 1, the image will be its original color.  
>   - Values in between create a partial grayscale effect.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+[!code-c[](./snippets/grayscaleeffect.fx?start=33&end=34)]
+
+This uses the [`dot`](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-dot) function to calculate the do product between the color's RGB values and the vector $(0.3, 0.59, 0.11)$.  This effectively calculates a weighted average where:
```
Spell checker FAIL :D

(Don't write tired)

calculate the "dot" product

AristurtleDev replied:
Fixed the spelling error

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+> [!NOTE]
+> These specific weights are based on how the human eye perceives brightness in different colors.  Green appears brighter to us than red, which appears brighter than blue.  
+>
+> The weighted values themselves are based on the formula that represents the luma component from the [ITU-R BT.601](https://en.wikipedia.org/wiki/Rec._601) standard, which is commonly used for converting RGB images to grayscale based on human perception where:
```
Now you are going just too far with this awesomeness :D There kids, take some MATH!

AristurtleDev replied:
![mindblown-tim-and-eric](https://github.com/user-attachments/assets/74ffb370-f6b5-4da8-b7cb-0d40d17ed30e)

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+    [!code-csharp[](./snippets/gamescene/gameover.cs?highlight=9-10)]
+
+5. Next, modify the `Update` method to handle the grayscale transition:
```
Would be nice to add a challenge for the user to also make this do the reverse once the game is unpaused :D

AristurtleDev replied:
I plan to go back in and add additional challenges once the overall tutorial is finialized.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
````diff
@@ -0,0 +1,392 @@
+
+```cs
+// Update the parameter value
+exampleEffect.Parameters["Saturation"].SetValue(0.5f);
````
Needs a call out to state we apply these values EVERY frame, as the shader simply applies effects based on the values it is sent as shaders are "normally" stateless, it does not remember what it was sent last time.

Kinda obvious but something that is not obvious to the beginner.

I also learned this is also how GPT's work as they are also stateless, you have to send the entire conversation each time you ask it another question.

AristurtleDev replied:
Added the following note

> [!NOTE]
> Notice how we set the shader parameters with the current saturation value every frame before beginning the sprite batch.  This is because shaders are stateless; they do not remember any values from the previous draw cycle.  Each time the GPU processes a shader, it only works with the parameters provided in that specific frame.  Event if the saturation value has not changed since the last frame, we still need to send it to the shader again to apply it.  This is why we update the shader parameters in the `Draw` method rather than only when the value is changed.

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+    spriteBatch.Begin();
+    spriteBatch.Draw(texture, position, color);
+    spriteBatch.End();
+
```
\Whitespace be gone

AristurtleDev replied:
Extra line removed

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
````diff
@@ -0,0 +1,392 @@
+
+    ```  
+
+3. Even though the effect to use is specified during the [**SpriteBatch.Begin**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch.Begin(Microsoft.Xna.Framework.Graphics.SpriteSortMode,Microsoft.Xna.Framework.Graphics.BlendState,Microsoft.Xna.Framework.Graphics.SamplerState,Microsoft.Xna.Framework.Graphics.DepthStencilState,Microsoft.Xna.Framework.Graphics.RasterizerState,Microsoft.Xna.Framework. ...
````
ONLY the last parameter value set is what is applied

AristurtleDev replied:
Updated and made it bold

> **only the last parameter value set**

SimonDarksideJ commented on articles/tutorials/building_2d_games/24_shaders/index.md:
````diff
@@ -0,0 +1,392 @@
+    spriteBatch.End();
+    ```
+
+4. There is an exception to #2 above.  In [Chapter 06: Working with Textures](../06_working_with_textures/index.md#layer-depth), we discussed the different [**SpriteSortMode**](xref:Microsoft.Xna.Framework.Graphics.SpriteSortMode) values that can be used when rendering.  From this chapter, we learned that when using [**SpriteSortMode.Immediate**](xref:Microsoft.Xna.Framework.Graphics.SpriteSortMo ...
````
Move the "However" to a new line so that it pops out, as it is kinda critical

AristurtleDev replied:
Moved it into an important admonition block

> [!IMPORTANT]
> As mentioned in [Chapter 06](../06_working_with_textures/index.md#layer-depth), [**SpriteSortMode.Immediate**](xref:Microsoft.Xna.Framework.Graphics.SpriteSortMode) can cause performance issues and should only be used when absolutely necessary.

SimonDarksideJ reviewed (changes requested):
Another fantastic chapter, with a few corrections/additions/suggestions.

Now considering a follow on Bounty for what other shader effects could be added to this tutorial later :D

TheKrumpet commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+For maximum compatibility, it's best to target the lower shader models:
+
+- For DirectX platforms: `vs_4_0_level_9_1` (vertex) and `ps_4_0_level_9_1` (pixel)
+- For OpenGL platforms: `vs_2_0` (vertex) and `ps_2_0` (pixel)
```
I'm not totally sure I agree with the callout for VS/PS 2.0 here, as they are pretty limited profiles compared to 3.0, and compatibility should be pretty much guaranteed at this point (it's 21 years old).

The shader template below also implements vs/ps 3.0.

AristurtleDev replied:
Thanks for pointing this out.  I can't remember why I stated it this way looking over my notes, but I've updated the section to be worded as follows

> For maximum compatibility, it's best to target the following shader models:
>
> - For DirectX platforms: `vs_4_0_level_9_1` (vertex) and `ps_4_0_level_9_1` (pixel)
> - For OpenGL platforms: `vs_3_0` (vertex) and `ps_3_0` (pixel)

TheKrumpet commented on articles/tutorials/building_2d_games/24_shaders/index.md:
```diff
@@ -0,0 +1,392 @@
+
+Different platforms support different shader capabilities, known as "shader models." When writing shaders for MonoGame, you need to consider compatibility across platforms. MonoGame supports the following shader models when targeting DirectX platforms:
+
+| Vertex Shader Model | Pixel Shader Model |
```
If we're being pendantic, then 'Shader Model' refers to the overall update to HLSL, and these are the _Shader Profiles_ you can use when compiling a particular shader function, so the headings here should be 'Vertex/Pixel Shader Profile'.

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-models

AristurtleDev replied:
Pendanticness is always welcome to ensure the correct terminology is used.  I have updated the headings to state "Profile" instead of "Model"

AristurtleDev committed:
Updated admonition based on feedback

AristurtleDev committed:
Add link for GLSL

AristurtleDev committed:
Updated paragraph based on feedback

AristurtleDev committed:
Add note about future shader support

AristurtleDev committed:
Add note about preprocessor directives

AristurtleDev committed:
Update samplers state list item based on feedback

AristurtleDev committed:
Added semantics link

AristurtleDev committed:
Add tip about ship model sample with blur shader

AristurtleDev committed:
Remove contraction "you've" to be "you have"

AristurtleDev committed:
Reword loading shader section to state that it is before our game implementation

AristurtleDev committed:
Capitalize G in MonoGame

AristurtleDev committed:
Reword section heading and title to be more clear that we're moving into implementation rather than theory

AristurtleDev committed:
Use unordered list

AristurtleDev committed:
Fixed the spelling issue
Don't tell me what to do, I'll write tutorials when I'm dead tired if I want to :P

AristurtleDev committed:
Add note about shaders being stateless

AristurtleDev committed:
Remove extra whitespace line

AristurtleDev committed:
Emphasis that only the last paramater value is set

AristurtleDev committed:
Move into admonition block to highlight importance

AristurtleDev committed:
Reword and suggest 3_0 for OpenGL models

AristurtleDev committed:
Use correct term "Profile" and not "Model"

AristurtleDev committed:
Add diagram and pipeline section

AristurtleDev committed:
Fix spelling issues

AristurtleDev committed:
Fix MonoGame spelling

AristurtleDev committed:
Merge branch '2d-tutorial-bounty' into 24-shaders
